src/app/course/search-course/search-course.component.ts

```diff
@@ -1,15 +1,25 @@
-import { Component, Output, EventEmitter } from "@angular/core";
+import { Component, Output, EventEmitter, ChangeDetectionStrategy } from "@angular/core";
+import { FormControl } from "@angular/forms";
+import { filter, debounce } from 'rxjs/operators';
+import { timer } from 'rxjs';
 
 @Component({
     selector: "app-search-course",
     templateUrl: "./search-course.component.html",
-    styleUrls: ["./search-course.component.less"]
+    styleUrls: ["./search-course.component.less"],
+    changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class SearchCourseComponent {
-    public search: string;
+    searchControl: FormControl;
     @Output() public courseSearched = new EventEmitter<string>();
 
-    public onCourseSearched() {
-        this.courseSearched.emit(this.search);
+    ngOnInit() {
+        this.searchControl = new FormControl("");
+        this.searchControl.valueChanges
+            .pipe(
+                filter((searchText: string) => searchText.length === 0 || searchText.length > 2),
+                debounce(() => timer(500))
+            )
+            .subscribe((searchInput: string) => this.courseSearched.emit(searchInput.trim()));
```

Comment:
instead of subscription you can do it this way: 
![image](https://user-images.githubusercontent.com/25720599/73425904-135e3680-4344-11ea-8c37-0baf590d4b5c.png)